src/potdata/samplers.py

```diff
@@ -616,3 +616,245 @@ def plot2(self, separation1: int, separation2: int, show: bool = False,
+
+        if "species" in soap_kwargs:
+            raise ValueError(
+                "The `species` argument in `soap_kwargs` is not allowed. "
+                "The species will be automatically determined from the structures."
+            )
+
+        soap = SOAP(species=species, **soap_kwargs)
+        atoms = [AseAtomsAdaptor.get_atoms(structure) for structure in data]
+        soap_vectors = list(soap.create(atoms))
+        
+        return soap_vectors
+
+    @staticmethod
+    def _select_by_species(
+        structures: list[Structure], vectors: list[np.ndarray], species: list[str]
+    ) -> list[np.ndarray]:
+        """Select SOAP vectors for atoms of specific species."""
+
+        def select_one(struct, vec):
+            indices = [i for i, s in enumerate(struct.species) if s.symbol in species]
+            return vec[indices]
+
+        selected = []
+        for struct, vec in zip(structures, vectors):
+            selected.append(select_one(struct, vec))
+
+        return selected
+
+    @staticmethod
+    def _get_soap_vector_structure(vectors: list[np.ndarray], pool: str) -> np.ndarray:
+        """Convert SOAP vectors of atoms to SOAP vectors of structures.
+
+        Return a 2D array of shape (n_structures, n_features).
+        """
+        if pool.lower() == "mean":
+            return np.asarray([np.mean(x, axis=0) for x in vectors])
+        elif pool.lower() == "concatenate":
+            return np.asarray([np.ravel(x) for x in vectors])
+        else:
+            supported = ["mean", "concatenate"]
+            raise ValueError(
+                f"Unsupported pooling method `{pool}`. Expected one of " f"{supported}."
+            )
+
+    @staticmethod
+    def _dim_reduction(x: np.ndarray, dim: int | float):
+        """Perform dimension reduction on a 2D array."""
+        pca = PCA(n_components=dim)
+        reduced = pca.fit_transform(x)
+
+        return reduced
+
+    def _cluster_kmeans(self, data: np.ndarray) -> list[int]:
+        """
+        Perform KMeans clustering and return cluster labels.
+        """
+
+        clustering = KMeans(n_clusters=self.kmeans_kwargs["n_clusters"], n_init=10).fit(data)
+        labels = clustering.labels_
+
+        unique_clusters = set(labels)
+        if len(unique_clusters) < self.kmeans_kwargs["n_clusters"]:
+            warnings.warn(
+                f"Number of clusters found ({len(unique_clusters)}) is less than "
+                f"specified n_clusters ({self.kmeans_kwargs['n_clusters']}). "
+                "Consider adjusting the parameters of KMeans."
+            )
+
+        return labels
+
+    def _select_kmeans(
+        self, data: list[Structure], cluster_labels: list[int], ratio: float
+    ) -> tuple[list[int], list[Structure]]:
+        """Select a subset of structures and indices for KMeans."""
+        unique_clusters = set(cluster_labels)
+        sample_size_per_cluster = int(ratio * len(unique_clusters))
```

Comment:
This line does not make sense. 

The `sample_size_per_cluster` should not depend on `unique_clusters`. It should be unique for each separate class. 

Let's do this. Let's make `ratio` multi-purpose:
- if it is an integer, a fixed number of `ratio` samples is chosen from each cluster. (of course, the line in 809 is needed)
- if it is a float number between 0 and 1, we chose a fraction of `ratio` for each cluster separately. For example, if `ratio=0.1', cluster 0 has 100 datapoints and cluster 1  has 20 data points, we will choose 10 and 2 data points from cluster 0 and 1, respectively.